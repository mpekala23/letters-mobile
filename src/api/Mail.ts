/* eslint-disable no-await-in-loop */
/* eslint-disable camelcase */
// The above is necessary because a lot of the responses from the server are forced snake case on us
import store from '@store';
import url from 'url';
import {
  Draft,
  Mail,
  TrackingEvent,
  MailTypes,
  MailStatus,
  Category,
  PostcardDesign,
  Image,
  Contact,
} from 'types';
import {
  addMail,
  setExistingMail,
  setActive,
  setContactsMail,
} from '@store/Mail/MailActions';
import { setUser } from '@store/User/UserActions';
import { popupAlert } from '@components/Alert/Alert.react';
import i18n from '@i18n';
import { addBusinessDays, differenceInHours } from 'date-fns';
import { estimateDelivery, getImageDims } from '@utils';
import { setCategories, setLastUpdated } from '@store/Category/CategoryActions';
import * as Sentry from 'sentry-expo';
import {
  updateContact,
  setActive as setActiveContact,
} from '@store/Contact/ContactActions';
import {
  getZipcode,
  fetchAuthenticated,
  API_URL,
  uploadImage,
  ApiResponse,
} from './Common';

interface RawTrackingEvent {
  id: number;
  name: string;
  date_modified: string;
  location: string;
}

interface RawImage {
  id: number;
  letter_id: number;
  created_at: string;
  updated_at: string;
  img_src: string;
}

interface RawMail {
  id: number;
  created_at: string;
  contact_id: number;
  content: string;
  sent: boolean;
  images: RawImage[];
  type: MailTypes;
  lob_status: string;
  last_lob_status_update: string;
  tracking_events?: RawTrackingEvent[];
  estimated_arrival: string;
  delivered: boolean;
}

function cleanLobStatus(status: string): MailStatus {
  // e.g 'In Transit' (single mail tracking event) and 'postcard.in_transit' (mail status)
  const normalizedStatus = status.split(' ').join('_').toLowerCase();
  if (!normalizedStatus) return MailStatus.Created;
  if (normalizedStatus.includes('in_transit')) return MailStatus.InTransit;
  if (normalizedStatus.includes('processed_for_delivery')) {
    return MailStatus.ProcessedForDelivery;
  }
  if (normalizedStatus.includes('in_local_area')) return MailStatus.InLocalArea;
  if (normalizedStatus.includes('mailed')) return MailStatus.Mailed;
  if (normalizedStatus.includes('returned_to_sender'))
    return MailStatus.ReturnedToSender;
  return MailStatus.Created;
}

async function cleanTrackingEvent(
  event: RawTrackingEvent
): Promise<TrackingEvent> {
  const location = event.location
    ? await getZipcode(event.location)
    : undefined;
  const date = new Date(event.date_modified);

  return {
    id: event.id,
    name: cleanLobStatus(event.name),
    location,
    date: new Date(date),
  };
}

export function mapTrackingEventsToMailStatus(
  events: TrackingEvent[]
): MailStatus {
  if (!events.length) return MailStatus.Created;
  const lastIdx = events.length - 1;
  return cleanLobStatus(events[lastIdx].name);
}

// cleans mail returned from getSingleMail
async function cleanMail(mail: RawMail): Promise<Mail> {
  const { type, content, id } = mail;
  const recipientId = mail.contact_id;
  let images: Image[] = [];
  if (mail.images.length) {
    try {
      images = await Promise.all(
        mail.images.map(async (rawImage) => {
          const dimensions = await getImageDims(rawImage.img_src);
          return {
            uri: rawImage.img_src,
            ...dimensions,
          };
        })
      );
    } catch {
      images = mail.images.map((rawImage) => ({ uri: rawImage.img_src }));
    }
  }
  const design = { image: images.length ? images[0] : { uri: '' } };

  const dateCreated = new Date(mail.created_at).toISOString();
  let status: MailStatus;
  let expectedDelivery = addBusinessDays(
    new Date(mail.created_at),
    6
  ).toISOString();
  const trackingEvents = !mail.tracking_events
    ? []
    : await Promise.all(
        mail.tracking_events.map((rawEvent) => cleanTrackingEvent(rawEvent))
      );
  if (!mail.sent) {
    status = MailStatus.Draft;
  } else if (mail.delivered) {
    status = MailStatus.Delivered;
  } else {
    status = mapTrackingEventsToMailStatus(trackingEvents);
    if (status === MailStatus.ProcessedForDelivery) {
      expectedDelivery = estimateDelivery(
        trackingEvents[trackingEvents.length - 1].date,
        status
      ).toISOString();
    }
  }

  if (mail.delivered) status = MailStatus.Delivered;
  if (type === MailTypes.Letter) {
    return {
      type,
      recipientId,
      content,
      id,
      status,
      dateCreated,
      expectedDelivery,
      images,
      trackingEvents,
    };
  }
  return {
    type,
    recipientId,
    content,
    id,
    status,
    dateCreated,
    expectedDelivery,
    design,
    trackingEvents,
  };
}

export async function getSingleMail(id: number): Promise<Mail> {
  const body = await fetchAuthenticated(url.resolve(API_URL, `letter/${id}`), {
    method: 'GET',
  });
  const data = body.data as RawMail;
  if (body.status !== 'OK' || !data) throw body;
  const cleanedMail = await cleanMail(data);
  return cleanedMail;
}

// cleans mail returned from getMail and defaults to getSingleMail if necessary
async function cleanMassMail(mail: RawMail): Promise<Mail> {
  if (!mail.lob_status || !mail.last_lob_status_update) {
    return getSingleMail(mail.id);
  }
  const { type, content, id } = mail;
  const recipientId = mail.contact_id;
  let images: Image[] = [];
  if (mail.images.length) {
    try {
      images = await Promise.all(
        mail.images.map(async (rawImage) => {
          const dimensions = await getImageDims(rawImage.img_src);
          return {
            uri: rawImage.img_src,
            ...dimensions,
          };
        })
      );
    } catch {
      images = mail.images.map((rawImage) => ({ uri: rawImage.img_src }));
    }
  }
  const design = { image: images.length ? images[0] : { uri: '' } };
  const dateCreated = new Date(mail.created_at).toISOString();

  let status;
  if (!mail.sent) {
    status = MailStatus.Draft;
  } else {
    status = mail.delivered
      ? MailStatus.Delivered
      : cleanLobStatus(mail.lob_status);
  }

  const expectedDelivery = estimateDelivery(
    new Date(mail.created_at),
    status
  ).toISOString();
  if (type === MailTypes.Letter) {
    return {
      type,
      recipientId,
      content,
      id,
      status,
      dateCreated,
      expectedDelivery,
      images,
    };
  }
  return {
    type,
    recipientId,
    content,
    id,
    status,
    dateCreated,
    expectedDelivery,
    design,
  };
}

export async function getMailByContact(
  contact: Contact,
  page = 1
): Promise<Mail[]> {
  const params = new URLSearchParams({ page: page.toString() });
  const body = await fetchAuthenticated(
    url.resolve(API_URL, `contact/${contact.id}/letters?${params}`),
    {
      method: 'GET',
    }
  );
  const data = body.data as {
    current_page: number;
    data: RawMail[];
    first_page_url: string;
    from: number;
    last_page: 1;
    last_page_url: string;
    next_page_url: string | null;
    to: number;
    total: number;
  };
  if (body.status !== 'OK' || !data || !data.data) throw body;
  // update the mail
  const clean = await Promise.all(
    data.data.map((raw) => {
      return cleanMassMail(raw);
    })
  );
  const existingMail = store.getState().mail.existing[contact.id];
  const existingArray = existingMail || [];
  const newMail = [...existingArray, ...clean];
  store.dispatch(setContactsMail(contact.id, page === 1 ? clean : newMail));
  // update the info in the contact
  const updatedContact = { ...contact };
  updatedContact.totalSent = data.total;
  updatedContact.mailPage = data.current_page + 1;
  updatedContact.hasNextPage = !!data.next_page_url;
  store.dispatch(updateContact(updatedContact));
  if (updatedContact.id === store.getState().contact.active.id) {
    store.dispatch(setActiveContact(updatedContact));
  }
  return clean;
}

export async function initMail(seedContacts?: Contact[]): Promise<void> {
  const contacts =
    seedContacts?.slice(0, 10) || store.getState().contact.existing;
  await Promise.all(
    contacts.map(async (contact) => {
      try {
        const pageOneMail = await getMailByContact(contact, 1);
        store.dispatch(setContactsMail(contact.id, pageOneMail));
      } catch (err) {
        Sentry.captureException(err);
      }
    })
  );
}

export async function getMail(page = 1): Promise<Record<string, Mail[]>> {
  const params = new URLSearchParams({ page: page.toString() });
  const body = await fetchAuthenticated(
    url.resolve(API_URL, `letters?${params}`),
    {
      method: 'GET',
    }
  );
  const data = body.data as { current_page: number; data: RawMail[] };
  if (body.status !== 'OK' || !data || !data.data) throw body;
  const newExisting: Record<string, Mail[]> = {};
  await Promise.all(
    data.data.map(async (raw) => {
      try {
        const clean = await cleanMassMail(raw);
        if (clean.recipientId in newExisting) {
          const mail = newExisting[clean.recipientId];
          let ix = 0;
          for (ix = 0; ix < mail.length; ix += 1) {
            const searchDate = mail[ix].dateCreated;
            if (
              clean.dateCreated &&
              searchDate &&
              clean.dateCreated > searchDate
            )
              break;
          }
          mail.splice(ix, 0, clean);
          newExisting[clean.recipientId] = mail;
        } else {
          newExisting[clean.recipientId] = [clean];
        }
      } catch (e) {
        Sentry.captureException(e);
        throw Error(e.message);
      }
    })
  );
  store.dispatch(setExistingMail(newExisting));
  return newExisting;
}

export async function getTrackingEvents(id: number): Promise<Mail> {
  const mail = await getSingleMail(id);
  const contactId = store.getState().contact.active.id;
  const currentMail = [...store.getState().mail.existing[contactId]];
  const ix = currentMail.findIndex((testMail) => testMail.id === id);
  currentMail[ix] = mail;
  store.dispatch(setActive(mail));
  store.dispatch(setContactsMail(contactId, currentMail));
  return mail;
}

export async function createMail(draft: Draft): Promise<Mail> {
  const { user } = store.getState().user;
  if (user.credit <= 0) {
    popupAlert({
      title: i18n.t('Compose.notEnoughCredits'),
      buttons: [{ text: i18n.t('Alert.okay') }],
    });
    throw new Error('Not enough credits');
  }

  const prepDraft = { ...draft };
  let imageExtension = {};
  if (prepDraft.type === MailTypes.Postcard) {
    try {
      if (prepDraft.design.custom) {
        prepDraft.design.image = await uploadImage(
          prepDraft.design.image,
          'letter'
        );
      }
      imageExtension = {
        s3_img_urls: [prepDraft.design.image.uri],
      };
    } catch (err) {
      Sentry.captureException(err);
      const uploadError: ApiResponse = {
        status: 'ERROR',
        message:
          err.message === 'timeout' ? 'Image upload timeout' : err.message,
        date: Date.now(),
        data: {},
      };
      throw uploadError;
    }
  } else if (prepDraft.images.length) {
    try {
      const uris = await Promise.all(
        prepDraft.images.map(async (image) => {
          const resultImage = await uploadImage(image, 'letter');
          return resultImage.uri;
        })
      );
      imageExtension = {
        s3_img_urls: uris,
      };
    } catch (err) {
      Sentry.captureException(err);
      const uploadError: ApiResponse = {
        status: 'ERROR',
        message:
          err.message === 'timeout' ? 'Image upload timeout' : err.message,
        date: Date.now(),
        data: {},
      };
      throw uploadError;
    }
  }

  const reqBody: Record<string, unknown> = {
    contact_id: prepDraft.recipientId,
    content: prepDraft.content,
    is_draft: false,
    type: prepDraft.type,
    size: prepDraft.type === MailTypes.Postcard ? prepDraft.size : undefined,
    ...imageExtension,
  };
  const body = await fetchAuthenticated(url.resolve(API_URL, 'letter'), {
    method: 'POST',
    body: JSON.stringify(reqBody),
  });
  if (body.status !== 'OK' || !body.data) throw body;

  const data = body.data as RawMail;
  const createdMail = await cleanMassMail(data);
  store.dispatch(addMail(createdMail));
  user.credit -= 1;
  store.dispatch(setUser(user));

  const contact = store.getState().contact.active;
  contact.totalSent += 1;
  store.dispatch(updateContact(contact));

  return createdMail;
}

interface RawCategory {
  created_at: string;
  id: 1;
  img_src: string;
  name: string;
  updated_at: string;
  blurb: string;
}

function cleanCategory(
  raw: RawCategory,
  subcategories: Record<string, PostcardDesign[]>
): Category {
  return {
    id: raw.id,
    name: raw.name,
    image: { uri: raw.img_src },
    blurb: raw.blurb,
    subcategories,
  };
}

interface RawDesign {
  id: number;
  created_at: string;
  updated_at: string;
  name: string;
  front_img_src: string;
  thumbnail_src: string;
  type: MailTypes;
  back: null;
  subcategory_id: number;
  designer?: string;
  content_researcher?: string;
}

async function cleanDesign(
  raw: RawDesign,
  categoryId?: number,
  subcategoryName?: string
): Promise<PostcardDesign> {
  try {
    const imageDims = await getImageDims(raw.front_img_src);
    const thumbnailDims = await getImageDims(raw.thumbnail_src);
    return {
      image: {
        uri: raw.front_img_src,
        ...imageDims,
      },
      thumbnail: { uri: raw.thumbnail_src, ...thumbnailDims },
      name: raw.name,
      id: raw.id,
      categoryId,
      subcategoryName,
      contentResearcher: raw.content_researcher,
      designer: raw.designer,
    };
  } catch (err) {
    return {
      image: {
        uri: raw.front_img_src,
      },
      thumbnail: { uri: raw.thumbnail_src },
      name: raw.name,
      id: raw.id,
      categoryId,
      subcategoryName,
      contentResearcher: raw.content_researcher,
      designer: raw.designer,
    };
  }
}

export async function getSubcategoriesById(
  categoryId: number
): Promise<Record<string, PostcardDesign[]>> {
  const body = await fetchAuthenticated(
    url.resolve(API_URL, `categories/${categoryId}/designs`),
    { method: 'GET' }
  );
  if (body.status !== 'OK' || !body.data) throw body;
  const data = body.data as Record<string, RawDesign[]>;
  const cleanData: Record<string, PostcardDesign[]> = {};
  const subNames = Object.keys(data);
  for (let ix = 0; ix < subNames.length; ix += 1) {
    const subName = subNames[ix];
    cleanData[subName] = await Promise.all(
      data[subName].map((raw: RawDesign) =>
        cleanDesign(raw, categoryId, subName)
      )
    );
  }
  return cleanData;
}

export async function getCategories(): Promise<Category[]> {
  try {
    const categoryState = store.getState().category;
    if (
      categoryState.lastUpdated &&
      differenceInHours(new Date(), new Date(categoryState.lastUpdated)) < 1 &&
      categoryState.categories.length
    ) {
      // if categories are loaded into the store and were refreshed less than
      // an hour ago, don't bother making this call
      return categoryState.categories;
    }
    const body = await fetchAuthenticated(url.resolve(API_URL, 'categories'));
    if (body.status !== 'OK' || !body.data) throw body;
    const data = body.data as RawCategory[];
    const categories: Category[] = await Promise.all(
      data.map(async (raw: RawCategory) => {
        const subcategories = await getSubcategoriesById(raw.id);
        return cleanCategory(raw, subcategories);
      })
    );
    const personalIx = categories.findIndex(
      (cat: Category) => cat.name === 'personal'
    );
    if (personalIx < 0 || !categories.length) {
      store.dispatch(setCategories([]));
      store.dispatch(setLastUpdated(null));
      return [];
    }
    const personalCategory = categories.splice(personalIx, 1);
    categories.unshift(personalCategory[0]);
    store.dispatch(setCategories(categories));
    store.dispatch(setLastUpdated(new Date().toISOString()));
    return categories;
  } catch (err) {
    Sentry.captureException(err);
    throw err;
  }
}
