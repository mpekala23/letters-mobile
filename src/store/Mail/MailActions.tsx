import { Draft, Mail, MailStatus, Image, PostcardDesign } from 'types';
import {
  SET_COMPOSING,
  SET_RECIPIENT_ID,
  SET_CONTENT,
  SET_IMAGES,
  SET_DESIGN,
  CLEAR_COMPOSING,
  ADD_MAIL,
  SET_ACTIVE,
  SET_STATUS,
  SET_DATE_CREATED,
  SET_EXPECTED_DELIVERY,
  SET_CONTACTS_MAIL,
  SET_EXISTING_MAIL,
  MailActionTypes,
} from './MailTypes';

export function setComposing(draft: Draft): MailActionTypes {
  return {
    type: SET_COMPOSING,
    payload: draft,
  };
}

export function setRecipientId(id: number): MailActionTypes {
  return {
    type: SET_RECIPIENT_ID,
    payload: id,
  };
}

export function setContent(content: string): MailActionTypes {
  return {
    type: SET_CONTENT,
    payload: content,
  };
}

export function setImages(images: Image[]): MailActionTypes {
  return {
    type: SET_IMAGES,
    payload: images,
  };
}

export function setDesign(design: PostcardDesign): MailActionTypes {
  return {
    type: SET_DESIGN,
    payload: design,
  };
}

export function clearComposing(): MailActionTypes {
  return {
    type: CLEAR_COMPOSING,
    payload: null,
  };
}

export function addMail(mail: Mail): MailActionTypes {
  return {
    type: ADD_MAIL,
    payload: mail,
  };
}

export function setActive(mail: Mail | null): MailActionTypes {
  return {
    type: SET_ACTIVE,
    payload: mail,
  };
}

export function setStatus(
  status: MailStatus,
  contactId: number,
  mailId: number
): MailActionTypes {
  return {
    type: SET_STATUS,
    payload: {
      contactId,
      mailId,
      status,
    },
  };
}

export function setDateCreated(
  dateCreated: Date,
  contactId: number,
  mailId: number
): MailActionTypes {
  return {
    type: SET_DATE_CREATED,
    payload: {
      contactId,
      mailId,
      dateCreated,
    },
  };
}

export function setExpectedDelivery(
  expectedDelivery: Date,
  contactId: number,
  mailId: number
): MailActionTypes {
  return {
    type: SET_EXPECTED_DELIVERY,
    payload: {
      contactId,
      mailId,
      expectedDelivery,
    },
  };
}

export function setContactsMail(
  contactId: number,
  mail: Mail[]
): MailActionTypes {
  return {
    type: SET_CONTACTS_MAIL,
    payload: {
      contactId,
      mail,
    },
  };
}

export function setExistingMail(mail: Record<string, Mail[]>): MailActionTypes {
  return {
    type: SET_EXISTING_MAIL,
    payload: mail,
  };
}
