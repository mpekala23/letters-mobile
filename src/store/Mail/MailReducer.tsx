import { MailTypes, MailStatus } from 'types';
import {
  SET_COMPOSING,
  SET_RECIPIENT_ID,
  SET_CONTENT,
  SET_IMAGE,
  SET_DESIGN,
  CLEAR_COMPOSING,
  ADD_MAIL,
  SET_ACTIVE,
  SET_STATUS,
  SET_DATE_CREATED,
  SET_EXPECTED_DELIVERY,
  SET_CONTACTS_MAIL,
  SET_EXISTING_MAIL,
  MailActionTypes,
  MailState,
} from './MailTypes';

const initialState: MailState = {
  composing: {
    type: MailTypes.Letter,
    recipientId: -1,
    content: '',
  },
  active: {
    id: -1,
    type: MailTypes.Letter,
    recipientId: -1,
    content: '',
    status: MailStatus.Created,
    dateCreated: new Date(),
    expectedDelivery: new Date(),
  },
  existing: {},
};

export default function LetterReducer(
  state = initialState,
  action: MailActionTypes
): MailState {
  const currentState = { ...state };
  switch (action.type) {
    case SET_COMPOSING:
      currentState.composing = action.payload;
      return currentState;
    case SET_RECIPIENT_ID:
      currentState.composing.recipientId = action.payload;
      return currentState;
    case SET_CONTENT:
      currentState.composing.content = action.payload;
      return currentState;
    case SET_IMAGE:
      if (currentState.composing.type !== MailTypes.Letter) return currentState;
      currentState.composing.image = action.payload;
      return currentState;
    case SET_DESIGN:
      if (currentState.composing.type !== MailTypes.Postcard)
        return currentState;
      currentState.composing.design = action.payload;
      return currentState;
    case CLEAR_COMPOSING:
      currentState.composing = {
        type: MailTypes.Letter,
        recipientId: -1,
        content: '',
      };
      return currentState;
    case ADD_MAIL:
      if (action.payload.recipientId in currentState.existing) {
        currentState.existing[action.payload.recipientId].unshift(
          action.payload
        );
      } else {
        currentState.existing[action.payload.recipientId] = [action.payload];
      }
      return currentState;
    case SET_ACTIVE:
      currentState.active = action.payload;
      return currentState;
    case SET_STATUS:
      if (!(action.payload.contactId in currentState.existing))
        return currentState;
      for (
        let ix = 0;
        ix < currentState.existing[action.payload.contactId].length;
        ix += 1
      ) {
        if (
          currentState.existing[action.payload.contactId][ix].id ===
          action.payload.mailId
        ) {
          currentState.existing[action.payload.contactId][ix].status =
            action.payload.status;
          break;
        }
      }
      return currentState;
    case SET_DATE_CREATED:
      if (!(action.payload.contactId in currentState.existing))
        return currentState;
      for (
        let ix = 0;
        ix < currentState.existing[action.payload.contactId].length;
        ix += 1
      ) {
        if (
          currentState.existing[action.payload.contactId][ix].id ===
          action.payload.mailId
        ) {
          currentState.existing[action.payload.contactId][ix].dateCreated =
            action.payload.dateCreated;
          break;
        }
      }
      return currentState;
    case SET_EXPECTED_DELIVERY:
      if (!(action.payload.contactId in currentState.existing))
        return currentState;
      for (
        let ix = 0;
        ix < currentState.existing[action.payload.contactId].length;
        ix += 1
      ) {
        if (
          currentState.existing[action.payload.contactId][ix].id ===
          action.payload.mailId
        ) {
          currentState.existing[action.payload.contactId][ix].expectedDelivery =
            action.payload.expectedDelivery;
          break;
        }
      }
      return currentState;
    case SET_CONTACTS_MAIL:
      currentState.existing[action.payload.contactId] = action.payload.mail;
      return currentState;
    case SET_EXISTING_MAIL:
      currentState.existing = action.payload;
      return currentState;
    default:
      return currentState;
  }
}
