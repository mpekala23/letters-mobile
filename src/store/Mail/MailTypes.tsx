import { Draft, Mail, MailStatus, Image, PostcardDesign } from 'types';

// Draft actions
export const SET_COMPOSING = 'mail/set_composing';
export const SET_RECIPIENT_ID = 'mail/set_recipient_id';
export const SET_CONTENT = 'mail/set_content';
export const SET_IMAGES = 'mail/set_images';
export const SET_DESIGN = 'mail/set_design';
export const CLEAR_COMPOSING = 'mail/clear_composing';

export interface MailState {
  composing: Draft;
  active: Mail | null;
  existing: Record<string, Mail[]>;
}

interface SetComposingAction {
  type: typeof SET_COMPOSING;
  payload: Draft;
}

interface SetRecipientIdAction {
  type: typeof SET_RECIPIENT_ID;
  payload: number;
}

interface SetContentAction {
  type: typeof SET_CONTENT;
  payload: string;
}

interface SetImagesAction {
  type: typeof SET_IMAGES;
  payload: Image[];
}

interface SetDesignAction {
  type: typeof SET_DESIGN;
  payload: PostcardDesign;
}

interface ClearComposingAction {
  type: typeof CLEAR_COMPOSING;
  payload: null;
}

// Mail actions
export const ADD_MAIL = 'mail/add_mail';
export const SET_ACTIVE = 'mail/set_active';
export const SET_STATUS = 'mail/set_status';
export const SET_DATE_CREATED = 'mail/set_date_created';
export const SET_EXPECTED_DELIVERY = 'mail/set_expected_delivery';
export const SET_CONTACTS_MAIL = 'mail/set_contacts_mail';
export const SET_EXISTING_MAIL = 'mail/set_existing_mail';

interface AddMailAction {
  type: typeof ADD_MAIL;
  payload: Mail;
}

interface SetActiveAction {
  type: typeof SET_ACTIVE;
  payload: Mail;
}

interface SetStatusAction {
  type: typeof SET_STATUS;
  payload: {
    contactId: number;
    mailId: number;
    status: MailStatus;
  };
}

interface SetDateCreatedAction {
  type: typeof SET_DATE_CREATED;
  payload: {
    contactId: number;
    mailId: number;
    dateCreated: Date;
  };
}

interface SetExpectedDeliveryAction {
  type: typeof SET_EXPECTED_DELIVERY;
  payload: {
    contactId: number;
    mailId: number;
    expectedDelivery: Date;
  };
}

interface SetContactsMailAction {
  type: typeof SET_CONTACTS_MAIL;
  payload: {
    contactId: number;
    mail: Mail[];
  };
}

interface SetExistingMailAction {
  type: typeof SET_EXISTING_MAIL;
  payload: Record<string, Mail[]>;
}

export type MailActionTypes =
  | SetComposingAction
  | SetRecipientIdAction
  | SetContentAction
  | SetImagesAction
  | SetDesignAction
  | ClearComposingAction
  | AddMailAction
  | SetActiveAction
  | SetStatusAction
  | SetDateCreatedAction
  | SetExpectedDeliveryAction
  | SetContactsMailAction
  | SetExistingMailAction;
