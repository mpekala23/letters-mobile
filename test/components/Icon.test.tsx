import * as React from 'react';
import { Icon } from '@components';
import { render, toJSON } from '@testing-library/react-native';
import Mail from '@assets/views/Report/MailLoud';

describe('Icon component', () => {
  it('should match snapshot', () => {
    const { container } = render(<Icon svg={Mail} />);
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should show the logo', () => {
    const { container } = render(<Icon svg={Mail} />);
    expect(container.children[0].props.xml).toEqual(Mail);
  });
});
