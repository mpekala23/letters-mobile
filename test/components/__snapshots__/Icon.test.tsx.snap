// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={84}
    bbWidth={103}
    fill="none"
    focusable={false}
    height={84}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {},
        Object {
          "flex": 0,
          "height": 84,
          "width": 103,
        },
      ]
    }
    vbHeight={84}
    vbWidth={103}
    width={103}
    xml="<svg width=\\"103\\" height=\\"84\\" viewBox=\\"0 0 103 84\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<path d=\\"M2 73.4357V38.8319C1.99981 37.5563 2.31448 36.3004 2.91609 35.1756C3.5177 34.0508 4.38765 33.0919 5.44875 32.3839L32.5738 14.3032C33.8473 13.4535 35.344 13 36.875 13C38.406 13 39.9027 13.4535 41.1762 14.3032L68.3013 32.3839C69.3624 33.0919 70.2323 34.0508 70.8339 35.1756C71.4355 36.3004 71.7502 37.5563 71.75 38.8319V73.4357M2 73.4357C2 75.4911 2.81652 77.4623 4.26992 78.9157C5.72333 80.3691 7.69457 81.1857 9.75 81.1857H64C66.0554 81.1857 68.0267 80.3691 69.4801 78.9157C70.9335 77.4623 71.75 75.4911 71.75 73.4357M2 73.4357L28.1562 55.9982M71.75 73.4357L45.5938 55.9982M28.1562 55.9982L2 38.5607M28.1562 55.9982L32.5738 58.9432C33.8473 59.7929 35.344 60.2463 36.875 60.2463C38.406 60.2463 39.9027 59.7929 41.1762 58.9432L45.5938 55.9982M45.5938 55.9982L71.75 38.5607\\" stroke=\\"#6D89B1\\" stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/>
<path d=\\"M73.4279 11.6465L76.6275 2.99998\\" stroke=\\"#6D89B1\\" stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/>
<path d=\\"M85 19L101 10\\" stroke=\\"#6D89B1\\" stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/>
<path d=\\"M87 32.858L96.0098 32.5846\\" stroke=\\"#6D89B1\\" stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/>
</svg>
"
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        Array [
          "fill",
        ]
      }
    >
      <RNSVGPath
        d="M2 73.4357V38.8319C1.99981 37.5563 2.31448 36.3004 2.91609 35.1756C3.5177 34.0508 4.38765 33.0919 5.44875 32.3839L32.5738 14.3032C33.8473 13.4535 35.344 13 36.875 13C38.406 13 39.9027 13.4535 41.1762 14.3032L68.3013 32.3839C69.3624 33.0919 70.2323 34.0508 70.8339 35.1756C71.4355 36.3004 71.7502 37.5563 71.75 38.8319V73.4357M2 73.4357C2 75.4911 2.81652 77.4623 4.26992 78.9157C5.72333 80.3691 7.69457 81.1857 9.75 81.1857H64C66.0554 81.1857 68.0267 80.3691 69.4801 78.9157C70.9335 77.4623 71.75 75.4911 71.75 73.4357M2 73.4357L28.1562 55.9982M71.75 73.4357L45.5938 55.9982M28.1562 55.9982L2 38.5607M28.1562 55.9982L32.5738 58.9432C33.8473 59.7929 35.344 60.2463 36.875 60.2463C38.406 60.2463 39.9027 59.7929 41.1762 58.9432L45.5938 55.9982M45.5938 55.9982L71.75 38.5607"
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={4285368753}
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={4}
      />
      <RNSVGPath
        d="M73.4279 11.6465L76.6275 2.99998"
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={4285368753}
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={4}
      />
      <RNSVGPath
        d="M85 19L101 10"
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={4285368753}
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={4}
      />
      <RNSVGPath
        d="M87 32.858L96.0098 32.5846"
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={4285368753}
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={4}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
